File: Question/47_Permutations2.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> cur;
        vector<bool> used(nums.size(), false);
        sort(nums.begin(), nums.end());
        dfs(nums, 0, used, cur, ans);
        return ans;
    }
    void dfs(vector<int>& nums, int level, vector<bool>& used, vector<int>& cur, vector<vector<int>>& ans) {
        if(level >= nums.size()) {
            ans.push_back(cur);
            return;
        }

        for(int i = 0; i < nums.size(); ++i) {
            if(used[i] == true) continue;
            cur.push_back(nums[i]);
            used[i] = true;
            dfs(nums, level + 1, used, cur, ans);
            used[i] = false;
            cur.pop_back();
            while(i < nums.size() - 1 && nums[i] == nums[i + 1])
                ++i;
        }
    }
};

File: Question_Java/47_Permutations2.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        dfs(nums, new boolean[nums.length], new ArrayList<Integer>(), ans);
        return ans;
    }

    private void dfs(int[] nums, boolean[] used, List<Integer> curAns, List<List<Integer>> ans) {
        if (curAns.size() == nums.length) {
            ans.add(new ArrayList<Integer>(curAns));
            return;
        }
        for (int i = 0; i < nums.length; ++i) {
            if (used[i])
                continue;
            used[i] = true;
            curAns.add(nums[i]);
            dfs(nums, used, curAns, ans);
            curAns.remove(curAns.size() - 1);
            used[i] = false;
            while (i < nums.length - 1 && nums[i] == nums[i + 1])
                ++i;
        }
    }
}